Mathlib/Dynamics/TopologicalEntropy/NetEntropy.lean

```diff
@@ -0,0 +1,408 @@
+the `limsup` definition as the default. Because of lemmas `coverEntropyInf_eq_iSup_netEntropyInfEnt`
+and `coverEntropy_eq_iSup_netEntropyEnt`, we make the same choice here. Theorems about the
+topological entropy of invariant subsets will be stated using only `coverEntropy`.
+
+## Main results
+- `coverEntropy_eq_iSup_netEntropyEnt`: equality between the notions of topological entropy defined
+with covers and with nets. Has a variant for `coverEntropyÌnf`.
+
+## Tags
+net, entropy
+
+## TODO
+Get versions of the topological entropy on (pseudo-e)metric spaces.
+-/
+
+namespace Dynamics
+
+open Set Uniformity UniformSpace
+
+variable {X : Type*}
+
+/-! ### Dynamical nets -/
+
+/-- Given a subset `F`, an entourage `U` and an integer `n`, a subset `s` of `F` is a
+`(U, n)`-dynamical net of `F` if no two orbits of length `n` of points in `s` shadow each other.-/
+def IsDynNetOf (T : X → X) (F : Set X) (U : Set (X × X)) (n : ℕ) (s : Set X) : Prop :=
+    s ⊆ F ∧ s.PairwiseDisjoint (fun x : X ↦ ball x (dynEntourage T U n))
+
+lemma IsDynNetOf.of_le {T : X → X} {F : Set X} {U : Set (X × X)} {m n : ℕ} (m_n : m ≤ n) {s : Set X}
+    (h : IsDynNetOf T F U m s) :
+    IsDynNetOf T F U n s :=
+  ⟨h.1, PairwiseDisjoint.mono h.2 (fun x ↦ ball_mono (dynEntourage_antitone T U m_n) x)⟩
+
+lemma IsDynNetOf.of_entourage_subset {T : X → X} {F : Set X} {U V : Set (X × X)} (U_V : U ⊆ V)
+    {n : ℕ} {s : Set X} (h : IsDynNetOf T F V n s) :
+    IsDynNetOf T F U n s :=
+  ⟨h.1, PairwiseDisjoint.mono h.2 (fun x ↦ ball_mono (dynEntourage_monotone T n U_V) x)⟩
+
+lemma isDynNetOf_empty {T : X → X} {F : Set X} {U : Set (X × X)} {n : ℕ} :
+    IsDynNetOf T F U n ∅ :=
+  ⟨empty_subset F, pairwise_empty _⟩
+
+lemma isDynNetOf_singleton (T : X → X) {F : Set X} (U : Set (X × X)) (n : ℕ) {x : X} (h : x ∈ F) :
+    IsDynNetOf T F U n {x} :=
+  ⟨singleton_subset_iff.2 h, pairwise_singleton x _⟩
+
+/-- Given an entourage `U` and a time `n`, a dynamical net has a smaller cardinal than a dynamical
+  cover. This lemma is the first of two key results to compare two versions topological entropy:
+  with cover and with nets, the second being `coverMincard_le_netMaxcard`.-/
+lemma IsDynNetOf.card_le_card_of_isDynCoverOf {T : X → X} {F : Set X} {U : Set (X × X)}
+    (U_symm : SymmetricRel U) {n : ℕ} {s t : Finset X} (hs : IsDynNetOf T F U n s)
+    (ht : IsDynCoverOf T F U n t) :
+    s.card ≤ t.card := by
+  have (x : X) (x_s : x ∈ s) : ∃ z ∈ t, x ∈ ball z (dynEntourage T U n) := by
+    specialize ht (hs.1 x_s)
+    simp only [Finset.coe_sort_coe, mem_iUnion, Subtype.exists, exists_prop] at ht
+    exact ht
+  choose! F s_t using this
+  simp only [mem_ball_symmetry (U_symm.dynEntourage T n)] at s_t
+  apply Finset.card_le_card_of_injOn F (fun x x_s ↦ (s_t x x_s).1)
+  exact fun x x_s y y_s Fx_Fy ↦
+    PairwiseDisjoint.elim_set hs.2 x_s y_s (F x) (s_t x x_s).2 (Fx_Fy ▸ (s_t y y_s).2)
+
+/-! ### Maximal cardinal of dynamical nets -/
```

Comment:
```suggestion
/-! ### Maximal cardinality of dynamical nets -/
```
here and below

Reply:
There was "cardinal" in other places too, but now they're all fixed. Thanks!